File: anagram/anagram.rb
class Anagram

  def initialize(word)
    @word = word.downcase
  end

  def match(words)
    words.select { |w| anagram?(w) }
  end

  private

  def anagram?(candidate)
    return false if @word == candidate.downcase

    @word.chars.sort == candidate.downcase.chars.sort
  end

end

File: elixir/anagram/anagram.exs
defmodule Anagram do
  @doc """
  Returns all candidates that are anagrams of, but not equal to, 'base'.
  """
  @spec match(String.t, [String.t]) :: [String.t]
  def match(base, candidates) do
    Enum.filter(candidates, &anagram?(String.downcase(base), String.downcase(&1)))
  end

  defp anagram?(base, base), do: false
  defp anagram?(base, candidate), do: do_sort(base) == do_sort(candidate)

  defp do_sort(candidate), do: String.graphemes(candidate) |> Enum.sort
end
